net: tc: Allow user to select pre-emptive threading for RX/TX

jukkar opened the pull request:
Let user to decide whether the RX/TX threads are run in either
co-operative or pre-emptive thread priority.
Default is co-operative threading.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: tc: Allow user to select pre-emptive threading for RX/TX

Let user to decide whether the RX/TX threads are run in either
co-operative or pre-emptive thread priority.
Default is co-operative threading.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
@armandciejak  we discussed this in networking forum. This PR is just experimenting and needs more testing before it is ready to upstream.

andrewboie commented:
What's the functional implication of this? In other words, why would an application want to enable this feature, or leave it disabled?

andrewboie commented:
If this is related to our conversation about SMP the other day, it's worth noting the co-op threads maintain their individual scheduling guarantees even in an SMP environment. It's just that if you implement a critical section between two threads, relying on one thread having a higher co-op priority than another to guarantee exclusion, that the other thread could be running simultaneously on another CPU, and exclusion guarantees fall over.

jukkar commented:
> What's the functional implication of this? In other words, why would an application want to enable this feature, or leave it disabled?

This was related to #27662 where incoming network traffic can starve the system. By running RX/TX threads in pre-emptive priority, some other high priority thread is allowed to run.

andrewboie commented:
> > What's the functional implication of this? In other words, why would an application want to enable this feature, or leave it disabled?
> 
> This was related to #27662 where incoming network traffic can starve the system. By running RX/TX threads in pre-emptive priority, some other high priority thread is allowed to run.

ahh understood thanks!

jukkar committed:
tests: net: Adjust the thread priorities

Adjust the test thread priority so that the test / IP stack
has a chance to run and the test passes.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
samples: net: Adjust the thread priorities

If networking pre-emptive thread priorities are enabled,
then use the proper macro to enable them.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: Adjust the thread priorities

If networking pre-emptive thread priorities are enabled,
then use the proper macro to enable them.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

rlubos commented on subsys/net/ip/Kconfig:
```diff
@@ -167,6 +167,26 @@ config NET_TC_RX_COUNT
+	help
+	  With co-operative threads, the thread cannot be pre-empted.
+
+config NET_TC_THREAD_PREEMPTIVE
```
Should this variant me marked `[EXPERIMENTAL]` perhaps for now?

jukkar replied:
> Should this variant me marked `[EXPERIMENTAL]` perhaps for now?

Yes, I will add that.

rlubos reviewed (commented):
I've tested a few samples with preemptive threads, `lwm2m_client` worked fine, same as `echo_client`/`echo_server` on OpenThread. But when I tried to run `echo_client/server` against each other on qemu_x86, I've ended up with the following:
![image](https://user-images.githubusercontent.com/22167799/95580356-7eb1e980-0a37-11eb-8524-aafed22eeb3a.png)
Looked like the client tried to send data before the TCP connection was complete. Not sure what's the root cause of it.

jukkar commented:
> But when I tried to run `echo_client/server` against each other on qemu_x86, I've ended up with the following

Thanks for the info, I did not try this combo in my testing but will check it next.

jukkar commented:
> Looked like the client tried to send data before the TCP connection was complete. Not sure what's the root cause of it.

The reason was that `main()` was running in preemptive priority 0, and network TX/RX thread in priority 8. After setting the main to run in priority 8, the connectivity was ok. This is yet another "confusion" point for the application developers, the thread priorities need to be selected carefully in order to let the certain threads run in correct time. We will probably see lot of issues created for this in the future.

jukkar commented:
Changes:
* adjust the main() priority to be 8 in echo-client

AntoineZen commented:
I d'like to jump in here. We are also hit by the issue, and by a case of "Looked like the client tried to send data before the TCP connection was complete". I was quite confuse by the following:

> This is yet another "confusion" point for the application developers, the thread priorities need to be selected carefully in order to let the certain threads run in correct time.

IMO, in a RTOS with preemptive task, this should not be necessary. They are synchronization primitive, and a thread must use them (encapsulated in the socket API) to wait for the network thread to be ready if required. In an RTOS this could involve priory inheritance. Changing thread priority to solve such problem seems a poor to me. Put a semaphore or another barrier to block the application until the connection is completed.

Also, application designer should be able to select the network thread priority depending on they requirement.

jukkar commented:
> Put a semaphore or another barrier to block the application until the connection is completed.

Sure, that is a good solution. I was just trying to get the existing networking samples to work with minimal changes to them.

jukkar commented:
> Also, application designer should be able to select the network thread priority depending on they requirement.

Yes, allowing user to select the exact thread priority is still missing here. The situation is a bit more complex as we can have max. 8 RX or TX threads, and one would need to select the base priority correctly. That is the reason why the base priority is set to 8 in the current commits.

jukkar committed:
net: tc: Refactor RX and TX thread priorities

Set the RX/TX thread priorities so that if cooperative priorities
are used, then lowest priority thread will have priority -1 which
is the lowest cooperative priority. The higest net thread priority
will depend on number of traffic classes but with max value 8,
the highest priority will be -8.

If preemptive priorities are used, then highest priority thread
will have priority 0, which is the highest preemptive priority.
In this case, the lowest thread priority will be 7 if there are
8 traffic classes.

The motivation for this change is that for cooperative priorities
we want to let other cooperative priority threads to run before
networking. But if preemptive priorities are used, we want
networking threads to run reasonably often compared to other
preemptive priority threads.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
* Fixed merge conflict
* Fixed thread priorities in samples and net
* Refactored the RX/TX thread priorities (see the commit `net: tc: Refactor RX and TX thread priorities` for details)

jukkar commented:
* Fixed the unit tests

jukkar commented:
Hi all! I am hoping we could get this one merged soonish so I am asking reviews some help to review this. Note that pre-emptive priority support for networking thread it is marked as experimental and is turned off by default. So there should be minimal risk here, and the sooner we merge this, the better the testing coverage by the users there will be.

rlubos reviewed (approved):
Echo samples with preemptive threads work fine now.

alexanderwachter reviewed (approved):
+1 for 6locan
